client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java

```diff
@@ -186,8 +191,143 @@ private Checkpoint completeCheckpoint(String checkpointName) {
+            val state = getState();
+            switch (state.getConfigState()) {
+                case INITIALIZING:
+                    doInit(state);
+                    break;
+                case REINITIALIZING:
+                    doReinit(state);
+                    break;
+                case READY:
+                    // This boolean will help know if the update actually succeeds or not.
+                    AtomicBoolean successfullyUpdated = new AtomicBoolean(true);
+                    long newGen = state.getGeneration() + 1;
+                    synchronizer.updateState((s, updates) -> {
+                        successfullyUpdated.set(s.equals(state));
+                        // If successfullyUpdated is false then that means the current state where this update should
+                        // take place (i.e. state with READY configState, etc.) is not the state we are in so we do not
+                        // make the update.
+                        if (successfullyUpdated.get()) {
+                            updates.add(new ReaderGroupStateResetStart(config, newGen));
+                        }
+                    });
+                    val newState = synchronizer.getState();
+                    // If successfullyUpdated is true then that means the update can take place and at this point the
+                    // update is completed and it is in the desired state on which doReinit is called to continue the
+                    // reset with this newState.
+                    if (successfullyUpdated.get()) {
+                        doReinit(newState);
+                        return;
+                    }
+                    break;
+                case DELETING:
+                    doDelete(state);
+                    throw new ReinitializationRequiredException("This ReaderGroup does not exist.");
+            }
+        }
+    }
+
+    public void createState(ReaderGroupConfig config) {
+        val state = getState();
+        if (state.getConfigState() == ReaderGroupState.ConfigState.INITIALIZING && state.getConfig().equals(config)) {
+            doInit(state);
+        }
+    }
+
+    public void deleteState() {
+        while (true) {
+            val state = getState();
+            switch (state.getConfigState()) {
+                case INITIALIZING:
+                    doInit(state);
+                    break;
+                case REINITIALIZING:
+                    doReinit(state);
+                    break;
+                case READY:
+                    if (updateConfigState(ReaderGroupState.ConfigState.DELETING, state)) {
+                        val newState = getState();
+                        doDelete(newState);
+                        return;
+                    }
+                    break;
+                case DELETING:
+                    doDelete(state);
+                    return;
+            }
+        }
+    }
+
+    public void doInit(ReaderGroupState state) {
```

Comment:
these methods should be private now - doInit, doReset, doDelete etc

Reply:
Made them private